Organize constants into version modules

stargaser opened the pull request:
Constants are organized into modules for different versions [#4958].

- Updated constants are available in modules ``codata2014`` (physical constants [#4109]), ``iau2015``, (astronomical constants [#4397]), and ``astropyconst20`` (together). These constants are available in the astropy.constants namespace. 

- The v1.3 constants are available in modules ``codata2010``,  ``iau2012``, and ``astropyconst13``.

- The G*M constants have been added for the Sun, Jupiter and the Earth [#4153].

- Tests have been updated in the ``constants``, ``units``, ``cosmology`` and ``modeling`` subpackages. (The CODATA 2014 updates to  Boltzmann's constant and the Stefan-Boltzmann constant affect detailed tests involving temperatures, and changes to L_sun affect the normalization of the blackbody.)

Here is the API that has been implemented:
```
# Import latest recommended constants
from astropy import constants as const

# Import constants for a specific Astropy version (needs a better name for class)
from astropy.constants import astropyconst20 as const

# Import constants from a specific set (not a full set of constants necessarily)
from astropy.constants import codata2010 as const

# In all cases, constants can then be used as e.g.
const.au
const.c
const.km

# and by changing the import, can then switch to a different set.
# People using the first import should expect changes over time.
# Existing users should switch to using astropyconst13 if they
# want to reproduce results produced with Astropy versions until
# now
```

- [x] Organize constants into different modules
- [x] Make the registry specific to each collection of constant
- [x] Update the default constants to CODATA2014 and IAU2015
- [x] Pass all existing tests
- [x] Update CHANGES.rst with this pull request number.
- [x] Add tests for the different types of imports and making sure the old versions of constants are self-consistent (e.g. ``e``, ``e_esu``, ``e.esu``, ``e.si``)
- [x] Check that the rather large changes in blackbody and cosmology tests are reasonable given the updates to values of constants.
---
edit by @bsipocz: this will close #4958 close #4153

stargaser commented:
This PR is now ready for review.

The references for constants are checked before issuing a warning about a constant already having a definition.

bsipocz commented on CHANGES.rst:
```diff
@@ -8,6 +8,12 @@ New Features
+    constants in the CODATA2014 class and astronomical constants in the
+    ASTRON_V2_0 class. These updated constants are available in the
+    ``astropy.constants`` namespace. The constants for astropy v1.3
+    are retained in separate version classes. [#4958, #4109, #4397]
```
Please add the number of this PR here. Also remove 4109 and 4397 as those PRs were closed without merge and you didn't bring over the code here.

stargaser replied:
Added 6083 for this pull request, and removed 4109 and 4397.

Also added another bullet for 4153.

bsipocz commented:
I haven't followed closely the discussions that lead to this PR, but find the ``ASTRON_*`` names very misleading as I associate it with [ASTRON](https://www.astron.nl/) (the Dutch radioastronomy institute). Would it be possible to call it something else, e.g. ``ASTRO_*`` by dropping the ``N``?

astrofrog commented:
What are the ``ASTRON_`` constants exactly? If they are IAU ones, why not just use ``IAU_``?

mhvk commented:
Did like the idea of giving each class its own registry. Really hope to look in more detail later...

pllim commented on astropy/constants/si.py:
```diff
@@ -6,157 +6,329 @@
-# Planck constant
-h = Constant('h', "Planck constant", 6.62606957e-34, 'J s', 0.00000029e-34,
-             'CODATA 2010', system='si')
+class CODATA2014:
```
I am not comfortable with using old-style class syntax for Python 2, while we still support Python 2.

stargaser commented:
@pllim The default version with latest values would be in the `astropy.constants` namespace. Copied here is the API we discussed at pyastro17:

```
# Import latest recommended constants
from astropy import constants as const

# Import constants for a specific Astropy version (needs a better name for class)
from astropy.constants import ASTROPYCONST20 as const

# Import constants from a specific set (not a full set of constants necessarily)
from astropy.constants import CODATA2010 as const

# In all cases, constants can then be used as e.g.
const.au
const.c
const.km

# and by changing the import, can then switch to a different set.
# People using the first import should expect changes over time.
# Existing users should switch to using ASTROPYCONST13 if they
# want to reproduce results produced with Astropy versions until
# now
```

stargaser commented:
Unfortunately I have realized that the registry in `Constant._registry` is still a problem, because we define constants twice with the same name, and `name.lower()` is the key to the registry dict.

Right now I just have the "latest" constants in the registry. So this problem occurs when using the old constants:

```
import CODATA2010 as const
const.e
# returns the CODATA2010 constant

const.e.si
# returns the value in the registry which is CODATA2014
const.e.emu
# returns the value in the registry which is CODATA2014
```
(in this case the name is `'Electron charge'` and the key to the registry is `'electron charge'`.

I haven't figured out how to move the registry to the CODATA2010, CODATA2014, etc. classes and still have the functionality work. The `Constant` and `ConstantMeta` classes both make use of the registry.

A hack-y solution is to make names different for different versions of the same constant, e.g. "Electron charge (CODATA2010)". I doubt this would be satisfactory.

stargaser commented:
On further consideration, would it be so bad to change the names of the "old" constants, by indicating the version in the name? It would make the current solution work.

mhvk commented:
@stargaser - in `units`, there is a concept of registries that can be set (as in `u.set_enabled_units`); this uses a list to implement a simple stack -- perhaps something similar could work here? I'll try to finally have a proper look at the PR today -- the basis is clearly there!

p.s. Whatever the precise solution, though, I'm quite convinced the different variants need their own registry...

mhvk commented on astropy/modeling/tests/test_blackbody.py:
```diff
@@ -32,8 +32,8 @@ def test_evaluate(self, temperature):
 
-        assert_quantity_allclose(b(1.4 * u.micron), 4756726.111003904 * u.Jy)
-        assert_quantity_allclose(b(214.13747 * u.THz), 4756726.111003904 * u.Jy)
+        assert_quantity_allclose(b(1.4 * u.micron), 4734464.498937388 * u.Jy)
```
This is a very large change. Is it obvious what changed?

stargaser replied:
The relative change is 4.7e-3 which is due to the same relative change in L_sun. It's because the blackbody is normalized to `bolometric_flux = 1000 * u.L_sun / (4 * np.pi * (1.5 * u.pc)**2)`.

mhvk replied:
That makes sense, but what then surprises me is that in the documentation table `L_sun` has not changed.

mhvk commented on astropy/cosmology/tests/test_cosmology.py:
```diff
@@ -1081,13 +1081,13 @@ def test_critical_density():
     tcos = core.FlatLambdaCDM(70.4, 0.272, Tcmb0=0.0)
     assert allclose(tcos.critical_density0,
-                    9.31000324385361e-30 * u.g / u.cm**3)
+                    9.309668456020899e-30 * u.g / u.cm**3)
```
For this also rather large change, is it mostly `G` that changed?

stargaser replied:
The relative change is 3.596e-5 which is exactly the relative change in G.

mhvk replied:
Ah, good that those changes are understood! It will be important to mention this in the what's new section.

mhvk commented:
Larger comment: the problems here happen because the `Constant` class keeps the registry. I think the solution would be to subclass `Constant` for each set, ie., something like
```
class CODATA2010(Constant):
    default_reference = 'CODATA 2010'
    # which should be used when no reference is passed in.

h = CODATA2010(...)
```

I would strongly suggest to put these definitions in separate files (or directories?) rather than separate classes in, e.g., `si.py`, as it seems obvious that a single file will at some point just become too large. 

I have not thought it through too far, but it might also make sense to have the actual containers be *instances* rather than classes, where they could be instances of something that sets everything up correctly when passed on the file/module (or multiple files, modules?).

Finally, instead of `ASTRON`, maybe just `IAU`?

p.s. I do need to think more about it, but it is nearing 2 AM here and my brain could use some sleep!

stargaser commented:
Thanks for the review! I'll try the subclassing. I think that within the class definition of `Constant`, I have to change `Constant._registry` to `cls._registry`; and in `ConstantMeta` I think I have to use `self._registry`.

mhvk commented:
@stargaser - yes, that should be sufficient. But maybe think about the other places -- e.g., any time ahead of a function/attribute, I'd definitely replace `Constant` with `self` or `cls` as appropriate. I think it also makes sense to replace it in `__repr__`.

By the way, I realized another reason to move to separate files is that, then, the old constants only get initialized if they are actually used.

stargaser commented:
Many thanks to @mhvk for suggestions. 

- Constants are in separate files / modules corresponding to version (CODATA2014, CODATA2010, IAU2012, IAU2015).
- There is a separate registry for each collection of constants
- The astronomical constants have versions of IAU2012 and IAU2015

Some tests of the prior versions of constants are still needed.

bsipocz commented on CHANGES.rst:
```diff
@@ -8,11 +8,16 @@ New Features
+
+  - The GM mass parameters recommended by IAU 2015 Resolution B 3 have been
+    added as ``GM_sun``, ``GM_jup``, and ``GM_earth``, for the Sun,
+    Jupiter and the Earth. [#4153]
```
nitpicking yet again, but could you list the PR number rather than (or in addition to) the issue number it implements?

stargaser commented:
I've done all the work I planned on this branch, and I believe I've addressed all the review comments. Let me know if it is okay to merge.

As far as making the old constants available -- even though a user can import and use the old constants, other subpackages such as ``units`` will still use the current constants. They reach into the ``constants`` and ``constants.si`` namespaces. I think this is set at import time.

mhvk commented on astropy/constants/CODATA2010.py:
```diff
@@ -0,0 +1,125 @@
+class EMCodata2010(Codata2010):
+    """An electromagnetic constant."""
+
+    @property
```
Here and elsewhere, you don't have to repeat the `cgs` property, just subclass from `EMConstant` as well, as in:
```
class EMCodata2010(Codata2010, EMConstant):
    pass
```

mhvk replied:
But I realized something: this `EMCodata2010` will have its own registry, which is not, I think, what we want. So, instead of `pass`, write
```
    _registry = Codata2010._registry
```

mhvk commented on astropy/constants/CODATA2010.py:
```diff
@@ -0,0 +1,125 @@
+    _registry = {}
+    _has_incompatible_units = set()
+
+    def __new__(cls, abbrev, name, value, unit, uncertainty,
```
I think rather than overwriting `__new__`, I would slightly adjust `Constant.__new__`, setting `reference=None` in the keyword list, and if it is `None`, changing it to `cls.default_reference` if that exists, i.e.,
```
if reference is None:
    reference = getattr(cls, 'default_reference', None)
    if reference is None:
        raise TypeError("{} requires a reference.".format(cls))
```

mhvk commented on astropy/constants/__init__.py:
```diff
@@ -25,9 +25,10 @@
 except ImportError:
     pass
 
-from .constant import Constant, EMConstant
```
I think we cannot change this import; people may count on `EMConstant` being available.

Similarly, I think you may need to keep the imports of `si` and `cgs` just so people can do `const.si.c`

mhvk commented on astropy/constants/constant.py:
```diff
@@ -89,15 +89,15 @@ class Constant(Quantity):
     def __new__(cls, abbrev, name, value, unit, uncertainty, reference,
```
Here is where I'd suggest the `reference=None` stuff from above.

mhvk commented on astropy/constants/constant.py:
```diff
@@ -89,15 +89,15 @@ class Constant(Quantity):
         # initialized here, and we store the unit in string form.
         inst = np.array(value).view(cls)
 
+        if (system in instances):
```
Why the parentheses around?

mhvk commented on astropy/constants/constant.py:
```diff
@@ -118,8 +118,8 @@ def __new__(cls, abbrev, name, value, unit, uncertainty, reference,
-        return ('<Constant name={0!r} value={1} uncertainty={2} unit={3!r} '
-                'reference={4!r}>'.format(self.name, self.value,
+        return ('<{0} name={1!r} value={2} uncertainty={3} unit={4!r} '
+                'reference={5!r}>'.format(self.__class__,self.name, self.value,
```
Missing space after `,`

mhvk commented on astropy/constants/constant.py:
```diff
@@ -191,7 +191,7 @@ def si(self):
         """
 
-        instances = Constant._registry[self.name.lower()]
+        instances = self.__class__._registry[self.name.lower()]
```
I don't think you need the `.__class__` here -- `self` has that link, no?

mhvk commented on astropy/constants/constant.py:
```diff
@@ -200,7 +200,7 @@ def cgs(self):
         """
 
-        instances = Constant._registry[self.name.lower()]
+        instances = self.__class__._registry[self.name.lower()]
```
idem.

mhvk commented on astropy/constants/tests/test_prior_version.py:
```diff
@@ -0,0 +1,157 @@
+    from ..CODATA2010 import h
+
+    # check that the value is fairly close to what it should be (not exactly
+    # checking because this might get updated in the future)
```
This should be exact, right, since `CODATA2010` can *not* be updated in the future! In fact, you might to explicitly check that it is *not* identical to `CODATA2014` (assuming that is the case)

mhvk commented on astropy/constants/tests/test_prior_version.py:
```diff
@@ -0,0 +1,157 @@
+    E = Q(100.00000348276221, 'V/m')
+
+    # e.cgs is too ambiguous and should not work at all
+    pytest.raises(TypeError, lambda: e.cgs * E)
```
I like to be a bit more explicit:
```
with pytest.raises(TypeError):
    e.cgs * E
```

mhvk commented on docs/units/equivalencies.rst:
```diff
@@ -217,7 +217,7 @@ here is an example::
     >>> omega_B = 2 * np.pi * beam_sigma**2
     >>> freq = 5 * u.GHz
     >>> u.Jy.to(u.K, equivalencies=u.brightness_temperature(omega_B, freq))
```
While you are at it, could you add `# doctest: +FLOAT_CMP` to the end of this line, and just insert the real answer below? I think the dots are only confusing to people reading the documentation...

mhvk commented:
Good point on that using a particular version of the constants does not guarantee that one's units are consistent with them. That slightly beats the point... The only way around that I can see is some configuration system. This should be possible, but I think it is not particularly easy, and would suggest that we keep that for another PR -- this one definitely lays the groundwork for it! Also, for me at least it is fantastic that we now have a way to ensure our constants are up to date again, and a mechanism to keep them up to date.

@pllim, @bsipocz, @eteq, @astrofrog - what do you think about this issue? (that, e.g., `u.M_sun` will always get the most recent version of the constant).

pllim commented:
> `u.M_sun` will always get the most recent version of the constant

Let's say I want to reproduce results from a paper calculated using this `u.M_sun`. How do I find out which one the author used? I am guessing the answer here is, "No matter what version of constants the author used, check the Astropy version." In that case, the author needs to document which Astropy version (and if it is dev, the git hash as well) in the paper. Is this a reasonable thing to expect from the author (the user)? If the answer is yes, then I have no objection.

bsipocz commented:
@mhvk - I'm OK if the units are using the most recent version of the constant, given that a user could overwrite it by using another set of constants. 

@pllim - I think that doesn't change much, that author still needed to state which version of the constants they used, stating the astropy version is literary an easier way to do it.

stargaser commented:
Changes requested by @mhvk are implemented. I don't understand the one about ``L_sun`` not showing the new value in the documentation table?

mhvk commented:
@stargaser - I was confused that you mentioned `L_sun` changed and that was why a test had to be adjusted, but you did not seem to have to adjust `L_sun` in the table of constants. But now that I look better, I see no table of constants was adjusted at all -- those get generated automatically after all (duh...); I got confused by the doctests that had changes in `sol_Rad`, etc. So, all OK, as far as I'm concerned.

@astrofrog, @eteq - I'd still like your OK, not so much about the code (which I think is great), but about now having a state where a user can import a specific set of constants, but that will *not* lead to associated units using those constants as well. I don't think that should stop this PR -- it is needed to make the constant/unit system configurable anyway and I think for 2.0 a warning that numbers have changed is OK -- but I want to be sure you agree.

eteq commented on astropy/constants/codata2010.py:
```diff
@@ -0,0 +1,113 @@
+
+    def __new__(cls, abbrev, name, value, unit, uncertainty,
+                reference=default_reference, system=None):
+        return(super(CODATA2010, cls).__new__(cls, abbrev, name, value, unit, uncertainty,
```
(very minor nitpick) - this line is over 80 chars - maybe line break after ``unit,`` instead of ``uncertainty,``?

stargaser commented:
@eteq good point about docs/constants/index.rst, I'll give it a shot right now

astrofrog commented on CHANGES.rst:
```diff
@@ -8,6 +8,17 @@ New Features
+    ``iau2015`` module. These updated constants are available in the
+    ``astropy.constants`` namespace, as well as the ``astropyconst20`` module.
+    The constants for astropy v1.3 are retained in separate version modules
+    ``codata2010``, ``iau2012`` and ``astropyconst13``. [#6083]
```
Small comment: I'm not sure if it makes sense to say that codata* and iau* are for specific Astropy versions. What about the following wording:

*Constants are now organized into version modules, with physical CODATA constants in the ``codata2010`` and ``codata2014`` sub-modules and astronomical constants defined by the IAU in the ``iau2012`` and ``iau2015`` sub-modules. The default constants in ``astropy.constants`` in Astropy 2.0 have been updated from ``iau2012`` to ``iau2015`` and from ``codata2010`` to ``codata2014``. The constants for 1.3 can be accessed in the ``astropyconst13`` sub-module and the constants for 2.0 (the default in ``astropy.constants``) can also be accessed in the ``astropyconst20`` sub-module [#6083]*

eteq commented:
@stargaser - Fine to make a new heading.  I'd call it something slightly more descriptive like "Collections of constants, including prior versions".  But definitely a different section is called for (and I don't *think* there's previous precedent for something like this).

stargaser commented:
@eteq @astrofrog -- changes made to `index.rst` and `CHANGES.rst`.

astrofrog commented on docs/constants/index.rst:
```diff
@@ -67,6 +67,58 @@ cannot be used in expressions without specifying a system::
+
+**Warning:** units such as ``u.M_sun`` will use the current version of the
+corresponding constant. When using prior versions of the constants,
+quantities should be constructed with constants instead of units.
```
There is a standard Sphinx directive for warnings: ``.. warn:`` (see elsewhere in docs) - can you switch to using this?

eteq replied:
On the theory that @stargaser is done for the day, I figured I'd just update this directly so we can merge.

eteq commented:
FYI, I fixed the last doc comment directly - @stargaser you might just want to have a quick look at that so you know what I did.

mhvk commented:
:fireworks: Thanks, @stargaser!
